src/main/scala/org/ergoplatform/mining/AutolykosPowScheme.scala

```diff
@@ -0,0 +1,212 @@
+package org.ergoplatform.mining
+
+import com.google.common.primitives.{Bytes, Ints, Longs}
+import org.ergoplatform.mining.difficulty.RequiredDifficulty
+import org.ergoplatform.modifiers.ErgoFullBlock
+import org.ergoplatform.modifiers.history._
+import org.ergoplatform.modifiers.mempool.ErgoTransaction
+import org.ergoplatform.nodeView.history.ErgoHistory
+import scorex.core.block.Block.Timestamp
+import scorex.crypto.authds.{ADDigest, SerializedAdProof}
+import scorex.crypto.hash.{Blake2b256, Blake2b512, Digest32}
+import scorex.util.{ModifierId, ScorexLogging}
+
+import scala.annotation.tailrec
+import scala.math.BigInt
+import scala.util.Try
+
+/**
+  * Autolykos PoW puzzle implementation.
+  * Mining process is implemented in inefficient way and should not be used in real environment.
+  *
+  * @see papers/yellow/pow/ErgoPow.tex for full description
+  * @param k - number of elements in one solution
+  * @param N - list size
+  */
+class AutolykosPowScheme(k: Int, N: Int) extends ScorexLogging {
+
+  assert(k <= 21, "k > 21 is not allowed due to genIndexes function")
+  assert(N <= 256 * 256 * 256, "N is not fitting into 3 byte unsigned integer")
+
+  /**
+    * Constant data to be added to hash function to increase it's calculation time
+    */
+  val M: Array[Byte] = (0 until 2048).toArray.flatMap(i => Longs.toByteArray(i))
+
+  /**
+    * Checks that `header` contains correct solution of the Autolykos PoW puzzle.
+    */
+  def validate(header: Header): Try[Unit] = Try {
+    val b = getB(header.nBits)
+    val msg = msgByHeader(header)
+    val s = header.powSolution
+
+    require(s.d < b || s.d > (q - b), s"Incorrect d=${s.d} for b=$b")
+    require(s.pk.getCurve == group.curve && !s.pk.isInfinity, "pk is incorrect")
+    require(s.w.getCurve == group.curve && !s.w.isInfinity, "w is incorrect")
+
+    val p1 = pkToBytes(s.pk)
+    val p2 = pkToBytes(s.w)
+    val f = genIndexes(Bytes.concat(msg, s.n)).map(ib => genElement(msg, p1, p2, Ints.toByteArray(ib))).sum.mod(q)
+    val left = s.w.multiply(f.bigInteger).add(s.pk.negate())
```

Comment:
better to add s.pk to the right thus saving one negate operation

Reply:
fixed